src/components/PlayField/PlayField.tsx

```diff
@@ -1,82 +1,54 @@
-import styles from "./PlayField.module.css";
-import Card from "../Card/Card";
 import React from "react";
-import { initialImages } from "../../constants/initialImages";
-import { IOpenCardsItem, IPlayingFieldItem } from "../../types/types";
-import { shuffleArray } from "../../utils/array";
+import Card from "../Card/Card";
+import { CardId } from "../../types/types";
+import styles from "./PlayField.module.css";
+import { getCardsById } from "./helpers";
 
-const images = shuffleArray([...initialImages, ...initialImages]);
-const initialCards: IPlayingFieldItem[] = images.map((url) => {
-  return {
-    url,
-    flip: false,
-    off: false,
-  };
-});
+const cardsById = getCardsById();
 const MAX_OPEN_CARDS = 2;
 
 function PlayField() {
-  const [openCards, setOpenCards] = React.useState<IOpenCardsItem[]>([]);
-  const [cards, setCards] = React.useState<IPlayingFieldItem[]>(initialCards);
-  const areBothCardsOpen = openCards.length === MAX_OPEN_CARDS;
+  const [openCardIds, setOpenCardIds] = React.useState<CardId[]>([]);
+  const [switchedCardIds, setSwitchedCardIds] = React.useState<CardId[]>([]);
+
   const checkResult = React.useCallback(() => {
-    const [cardOne, cardTwo] = openCards;
-    const newCards = [...cards];
-    const newCardOne = newCards[cardOne.index];
-    const newCardTwo = newCards[cardTwo.index];
-    const isSuccess = cardOne.url === cardTwo.url;
+    if (openCardIds.length !== MAX_OPEN_CARDS) {
+      return;
+    }
+
+    const openCardOne = cardsById[openCardIds[0]];
+    const openCardTwo = cardsById[openCardIds[1]];
 
-    if (isSuccess) {
-      newCardOne.off = true;
-      newCardTwo.off = true;
-    } else {
-      newCardOne.flip = false;
-      newCardTwo.flip = false;
+    if (openCardOne.imageUrl === openCardTwo.imageUrl) {
+      setSwitchedCardIds(switchedCardIds.concat(openCardIds));
     }
 
-    setCards(newCards);
-    setOpenCards([]);
-  }, [cards, openCards]);
+    setOpenCardIds([]);
+  }, [openCardIds, switchedCardIds]);
+
   React.useEffect(() => {
-    setTimeout(() => {
-      if (areBothCardsOpen) {
-        checkResult();
-      }
-    }, 1000);
-  }, [areBothCardsOpen, checkResult, openCards]);
+    setTimeout(checkResult, 1000);
+  }, [checkResult]);
 
-  const handleClick = (index: number, url: string) => {
-    if (areBothCardsOpen) {
-      return;
-    }
-    const card = cards[index];
-    if (card.flip) {
-      return;
-    }
-    const newCards = [...cards];
-    newCards[index] = { ...card, flip: true };
-    setCards(newCards);
```

Comment:
Now you don't need to keep all cards in state, and keep them updated on every change. And you can avoid matching a card by its index in the cards array